### PR TITLE
feat: managing preferences

### PR DESCRIPTION
**Pull request by emilyjablonski**
# Pull Request Template

## Issue Overview

This PR addresses #2748 #2750

[Figma design](https://www.figma.com/file/Y8bFd0aSz7eYJ2TCNid6s3/bloom-partners?node-id=5725%3A182890)

- [x] This change addresses the issue in full

## Description

This PR implements the two issues above, which allow you to manage preferences in the new `Settings` panel in partners.
If you login as an admin or a jurisdictional admin, you should be able to see the preferences from the jurisdictions you have access to.

From the main page you can add a preference, duplicate a preference, or edit a preference.

Note some differences from the design that were decided in Slack over the last couple of weeks:
* We added the "Is this option multi-select or exclusive?" question on an individual option for preferences like `Housing Authority Project-Based Voucher - Alameda` in which there is the generic opt out value, but also a custom one (`Do not consider me`). Since `Do not consider me` is also a "pass" option, it needs to behave exclusively (it cannot be checked with any other checkbox selected). If all options are checked as exclusive, the question renders as a radio.
* We added the "Can you opt out?" question to a preference, for preferences like `Housing Authority Project-Based Voucher - Alameda` - there are two optional items you can select, but you don't have to select anything.
* We removed the `Subtitle` field visually from the form under the assumption it is not often used, but it may be added back in the future. The data is maintained, but it is hidden from the listing view, application, and preferences form.
* We removed the delete functionality for a preference over concerns around how we might handle existing listings that are tied to those preferences, and may add it back in the future.
* We added a link to the root preference in the design since it is in the existing data model.
* The data model supports multiple jurisdictions, we are starting with just one.
* The data model supports multiple links, we are starting with just one.

## How Can This Be Tested/Reviewed?

Note that the vast majority of this experience will be as a jurisdictional admin, so I'd recommend doing most of your testing there, but a super admin will still have access to all preferences across jurisdictions.

Compare visually with the [Figma design](https://www.figma.com/file/Y8bFd0aSz7eYJ2TCNid6s3/bloom-partners?node-id=5725%3A182890) noting the differences from the design above.

Ensure the preferences populate according to your user.
Ensure you can add a preference with options, and exercise all the fields.
Ensure you can edit a preference / option.
Ensure you can duplicate a preference / option.
Ensure you can delete an option.
Ensure you can reorder options.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have added QA notes to the issue with applicable URLs
- [x] I have performed a self-review of my own code
- [x] I have reviewed the changes in a desktop view
- [x] I have reviewed the changes in a mobile view
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have assigned reviewers
- [ ] I have run `yarn generate:client` and/or created a migration if I made backend changes that require them
- [ ] I have exported any new pieces added to ui-components
- [x] My commit message(s) is/are polished, and any breaking changes are indicated in the message and are well-described
- [ ] Commits made across packages purposefully have the same commit message/version change, else are separated into different commits

## Reviewer Notes:

Steps to review a PR:

- Read and understand the issue, and ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Also review the acceptance criteria on the Netlify deploy preview (noting that these do not yet include any backend changes made in the PR)
- Either explicitly ask a clarifying question, request changes, or approve the PR if there are small remaining changes but the PR is otherwise good to go

## On Merge:

If you have one commit and message, squash. If you need each message to be applied, rebase and merge.

### CODE REVIEW

**Comment by emilyjablonski** [comment]
Remaining question for product/design: We currently have a strange UX around the ordering of rows, and where rows appear when you edit them, and add them. Touching base offline.

**Comment by emilyjablonski** [comment]
Same issue with AccessLint - it was failing and is now fixed, so it stays pending forever

File: sites/partners/src/settings/PreferenceDrawer.tsx [hunk 1]
```diff
@@ -276,7 +539,45 @@ const PreferenceDrawer = ({ drawerOpen, setDrawerOpen }: PreferenceDrawerProps)
...
+              setError("optionTitle", { message: t("errors.requiredFieldError") })
+              return
+            }
+            const existingOptionData = questionData?.options?.find(
+              (option) => optionData?.ordinal === option.ordinal
+            )
+
+            const getNewOrdinal = () => {
+              if (existingOptionData) return existingOptionData.ordinal
+              return questionData?.options?.length ? questionData?.options.length + 1 : 1
```
Comment (ludtkemorgan) [inline sites/partners/src/settings/PreferenceDrawer.tsx:555]: This isn't necessarily more readable, but you should be able to get rid of the ternary operator with 
```
return  (questionData?.options?.length || 0) + 1

File: sites/partners/src/settings/PreferenceDrawer.tsx [hunk 2]
```diff
@@ -112,44 +180,149 @@ const PreferenceDrawer = ({ drawerOpen, setDrawerOpen }: PreferenceDrawerProps)
...
+                    defaultChecked: questionData === null || !questionData?.hideFromListing,
+                    dataTestId: "show-on-listing-question-yes",
+                  },
+                  {
+                    id: YesNoAnswer.No,
+                    label: t("t.no"),
+                    value: YesNoAnswer.No,
+                    defaultChecked: questionData?.hideFromListing,
+                    dataTestId: "show-on-listing-question-no",
+                  },
```
Comment (seanmalbert) [inline sites/partners/src/settings/PreferenceDrawer.tsx:324]: When I try and update showOnListingQuestion it's updating optOutText instead, so I'm not able to update showOnListingQuestion.

File: sites/partners/pages/settings/index.tsx [hunk 3]
```diff
@@ -77,20 +107,38 @@ const Settings = () => {
...
+                if (aChar === bChar)
+                  return a.updatedAt > b.updatedAt ? -1 : a.updatedAt < b.updatedAt ? 1 : 0
+                return aChar < bChar ? -1 : aChar > bChar ? 1 : 0
+              })
+              .map((preference) => {
+                return {
+                  name: { content: preference?.text },
+                  jurisdiction: {
+                    content: preference?.jurisdictions?.reduce((acc, item, index) => {
+                      return `${acc}${index > 0 ? ", " : ""}${item.name}`
```
[the remark this answers is not in the document]
Comment (emilyjablonski) [reply sites/partners/pages/settings/index.tsx:123]: At the moment we don't allow preferences to be assigned to more than one jurisdiction through the form even though the data model supports that - but I believe there is an intention to do so, and so there is some existing support for multiple remaining from what we had before

File: sites/partners/src/settings/PreferenceDrawer.tsx [hunk 4]
```diff
@@ -253,20 +477,51 @@ const PreferenceDrawer = ({ drawerOpen, setDrawerOpen }: PreferenceDrawerProps)
...
                   dataTestId={"preference-option-link-title"}
+                  defaultValue={optionData?.links?.length > 0 ? optionData?.links[0].title : ""}
                 />
               </ViewItem>
             </GridCell>
           </GridSection>
           <GridSection columns={3} className={"mt-8"}>
             <GridCell>
-              <Field
+              <Field //TODO: Add "exclusive"
```
[the remark this answers is not in the document]
Comment (emilyjablonski) [reply sites/partners/src/settings/PreferenceDrawer.tsx:500]: That one's done, thank you! 🧹

File: sites/partners/src/settings/PreferenceDrawer.tsx (hunk 2 above)
Comment (emilyjablonski) [reply sites/partners/src/settings/PreferenceDrawer.tsx]: Fixed!

File: backend/core/src/seeder/seeds/listings/shared.ts [hunk 5]
```diff
@@ -242,7 +242,21 @@ export const liveWorkPreference: MultiselectQuestionSeedType = {
...
+      links: [
+        {
+          title: "Link Title",
+          url: "https://www.example.com",
+        },
+        {
+          title: "Link Title",
+          url: "https://www.example.com",
+        },
+      ],
```
Comment (seanmalbert) [inline backend/core/src/seeder/seeds/listings/shared.ts:258]: Can you update the title of one of these, it made me think there was a bug when reviewing the frontend.

**Review by seanmalbert — APPROVED** [review]
Looks good! The only remaining issue I'm running into, which wasn't introduced in this PR (and may be existing on prod), is that when a preference has opt out enabled and an exclusive option. If a user selects opt out and then selects the exclusive option, then the opt out option doesn't get selected.

**Comment by seanmalbert** [comment]
exclusive option fix looks good

**Review by ludtkemorgan — APPROVED** [review]

**Pull request merged**